Octo Ottano

robertlong13 opened the pull request.

robertlong13 commented:
Some substantial changes to our build script. This is still very much a quick-and-dirty WIP, but this is the basic idea.

I think I am actually going to undo this "extra_hwdef" handling and just actually make a "CubeOrange-Carbonix" and "CubeOrangePlus-Carbonix" board type, but it's late and I want to push what I have now. That will make things much more intuitive for outsiders/future-me. I got too complex with that bit, especially considering that every aircraft configuration uses the same hwdef customizations for now.

If we ever had a reason to make per-aircraft-config tweaks to hwdef, I could revisit that, but I don't think we will.

robertlong13 committed:
hwdef: merge hwdefs for Volanti and Ottano

robertlong13 committed:
Tools: specify defaults in aircraft_config xmls

robertlong13 committed:
hwdef: add Octo config for Ottano

This also required some shifting of servo stuff out of Common, so all
param files have been touched.

robertlong13 commented:
I have manually confirmed that this results in no change on the existing Volanti_AC_3 and Ottano_AC_2 builds from CI. Steps:
- Download defaults.parm over MAVFTP
- Sort the lines (there is a command to do this in VSCode; hit F1)
    - This step is needed because the shifting that occurred in the common params, this changes the order of the processed params
- Compare
- No changes other than comments

loki077 commented:
@robertlong13 I am good with the structure and also aircraft_config.py mods.

I will review further the params and everything.

I still think to simplify and keeping the structure
Common
--> Volanit
--> OttanoQuad
--> OttanoOcto

is good enough and there is no need for Ottano common as there a possibility in future Octo could grow to become more different than it is from quad ottano.

But I can accept as it is for now.

robertlong13 commented:
> is good enough and there is no need for Ottano common as there a possibility in future Octo could grow to become more different than it is from quad ottano.

Right now, they have more in common than they have different. One nice thing about this breakdown is that, if they start to grow apart, it will be reflected in the shrinking of that common param file, at which point it will be easy to make the judgement call to drop it.

At the end of the day, this setup makes my job easier

loki077 commented:
> > is good enough and there is no need for Ottano common as there a possibility in future Octo could grow to become more different than it is from quad ottano.
> 
> Right now, they have more in common than they have different. One nice thing about this breakdown is that, if they start to grow apart, it will be reflected in the shrinking of that common param file, at which point it will be easy to make the judgement call to drop it.
> 
> At the end of the day, this setup makes my job easier

@robertlong13 that make sense.

robertlong13 commented:
Remaining TODO:
- [x] Fill out missing parts of the xml
- [x] CxBIT handling for octo

robertlong13 committed:
CarbonixCommon: OttanoOcto: fix BATT4/BATT5

robertlong13 committed:
CarbonixCommon: bump OttanoOcto speeds

Increased all speeds by 10% to account for additional weight

robertlong13 committed:
CarbonixCommon: OttanoOcto GPS node IDs

robertlong13 committed:
CarbonixCommon: OttanoOcto: INS_ACCEL_FILTER to 20

This can be up at the higher end with our extensive notch filtering

robertlong13 committed:
CarbonixCommon: OttanoOcto: use stock notch atten.

robertlong13 committed:
CarbonixCommon: Ottano: correct sensor positions

robertlong13 committed:
Tools: devid scripts change "UAVCAN" to "DroneCAN"

robertlong13 committed:
Ottano: fix Octo compasses

These use different node IDs with this CAN architecture

robertlong13 committed:
OttanoOcto: VTOL tune

robertlong13 committed:
Ottano: enable VTOL pusher assist

robertlong13 committed:
OttanoOcto: VTOL pos/loit tune

robertlong13 committed:
OttanoCommon: remove ICE RC channel

robertlong13 committed:
sitl_engine: remove servo3 assumption for throttle

robertlong13 committed:
bit_esc: detect quad vs octo

robertlong13 committed:
Tools: add cx_prep_sitl

Allows for some clever launch.json stuff to launch particular Carbonix
SITL frames defined in vehicleinfo.py. Now that our SITL relies on a
complex collection of scripts, this is becoming necessary.

robertlong13 committed:
CarbonixCommon: add ottano realflight

robertlong13 committed:
Tools: aircraft_config: add --upload option

Useful for quickly flashing onto an autopilot for testing/debugging

robertlong13 committed:
Ottano: add batt8 fuel pressure sensor

robertlong13 committed:
Ottano: fix throttle linearization

robertlong13 committed:
Ottano: fix fuel burn reporting

robertlong13 committed:
OttanoOcto: swap GPS1/2

robertlong13 committed:
OttanoOcto: allow ESC notch down to 20Hz

The Octo spins much slower, especially midway through transition.
We need this.

robertlong13 committed:
OttanoOcto: increase Q_TRANS_FAIL

robertlong13 committed:
OttanoOcto: decrease Q_TRANSITION_MS

The Octo has a pitchup at the end of the transition, caused by odd wake
interaction between the VTOL layers. Shortening the time it takes to
spool down the VTOLs helps somewhat.

robertlong13 committed:
Ottano: increase weathervaning gain

robertlong13 committed:
Ottano: new EIB revision

Engine ignition and starter are controlled from a DroneCAN relay, and
12V/28V rail monitoring is added.

robertlong13 committed:
Ottano: add differential aileron

robertlong13 committed:
Ottano: fixed-wing roll tune

robertlong13 committed:
OttanoOcto: TECS tune

robertlong13 committed:
Ottano: fixed-wing tighten outer pitch loop

This was designed for Octo, but benefits all.

robertlong13 committed:
Ottano: increase NAVL1_DAMPING

LachlanConn reviewed (commented):
Some suggestions maybe 

- Swap compass ordering to be in lnline with bus 1 devices as primary and bus 2 devices as secondary
- Maybe make differential aileron change as common and could even squash into RLL_TUNE commit
- Comment on Pitch RMAX to say reason was too many G's

robertlong13 committed:
Ottano: fixed-wing rudder tune

Some of these aggressive original values were an attempt to make up for
the lack of differential aileron.

robertlong13 committed:
CarbonixCommon: ignore CAN ESC offset on receive

This is needed for the Octo, but it's a non-functional change for the
quads if I add it to common like this, as the offset is 0 on the quads.

robertlong13 committed:
sitl_engine: use RPM2

RealFlight as a sitl frame (at least in 4.5) can't report pusher RPM on
anything but RPM2.

Headless fakes its own RPM anyway, so this doesn't matter for it.

robertlong13 committed:
cpn_params: add Ottano Fuse1 and Fuse2

robertlong13 committed:
Tools: cx build: strip comments in periph defaults

These comments in the embedded defaults take up precious space, and they
are usually just lengthy explanations for DISABLE_CHECKS, which don't
need to be embedded in the firmware binary.

loki077 commented on libraries/AP_HAL_ChibiOS/hwdef/CarbonixCommon/aircraft_configuration/Ottano_AC_3.xml:
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0"?>
```
```
<?xml version="1.0"?>

<data>
    <aircraft>
        <model>Ottano</model>
        <model_version>AC_3</model_version>
        <model_description>
            <!-- Flight Controller: -->
            CubeOrangePlus,
            <!-- TODO -->
        </model_description>
        <status>active</status>
        <configuration_version>1</configuration_version>
    </aircraft>
    <flight_controller>
        <board_name>CubeOrangePlus-CX</board_name>
        <firmware_commit>cx_pilot_commit_id</firmware_commit>
        <firmware_path>arduplane.apj</firmware_path>
        <defaults_file>OttanoOcto.parm</defaults_file>
    </flight_controller>
    <cpn_list>
        <cpn id="11">
            <bus>1</bus>
            <board_name>Ottano-M1_CarbonixF405</board_name>
            <firmware_commit>cx_pilot_commit_id</firmware_commit>
            <firmware_path>AP_Periph.bin</firmware_path>
        </cpn>
        <cpn id="12">
            <bus>1</bus>
            <board_name>Ottano-M2_CarbonixF405</board_name>
            <firmware_commit>cx_pilot_commit_id</firmware_commit>
            <firmware_path>AP_Periph.bin</firmware_path>
        </cpn>
        <cpn id="13">
            <bus>1</bus>
            <board_name>Ottano-M3_CarbonixF405</board_name>
            <firmware_commit>cx_pilot_commit_id</firmware_commit>
            <firmware_path>AP_Periph.bin</firmware_path>
        </cpn>
        <cpn id="14">
            <bus>2</bus>
            <board_name>Ottano-M4_CarbonixF405</board_name>
            <firmware_commit>cx_pilot_commit_id</firmware_commit>
            <firmware_path>AP_Periph.bin</firmware_path>
        </cpn>
        <cpn id="25">
            <bus>2</bus>
            <board_name>Ottano-M5_CarbonixF405</board_name>
            <firmware_commit>cx_pilot_commit_id</firmware_commit>
            <firmware_path>AP_Periph.bin</firmware_path>
        </cpn>
        <cpn id="26">
            <bus>2</bus>
            <board_name>Ottano-M6_CarbonixF405</board_name>
            <firmware_commit>cx_pilot_commit_id</firmware_commit>
            <firmware_path>AP_Periph.bin</firmware_path>
        </cpn>
        <cpn id="27">
            <bus>2</bus>
            <board_name>Ottano-M7_CarbonixF405</board_name>
            <firmware_commit>cx_pilot_commit_id</firmware_commit>
            <firmware_path>AP_Periph.bin</firmware_path>
        </cpn>
        <cpn id="28">
            <bus>2</bus>
            <board_name>Ottano-M8_CarbonixF405</board_name>
            <firmware_commit>cx_pilot_commit_id</firmware_commit>
            <firmware_path>AP_Periph.bin</firmware_path>
        </cpn>
        <cpn id="41">
            <bus>2</bus>
            <board_name>servo_.MKS-HBL6625:010</board_name>
            <firmware_commit>b120233</firmware_commit>
        </cpn>
        <cpn id="42">
            <bus>2</bus>
            <board_name>servo_.MKS-HBL6625:006</board_name>
            <firmware_commit>b120233</firmware_commit>
        </cpn>
        <cpn id="43">
            <bus>2</bus>
            <board_name>servo_.MKS-HBL6625:011</board_name>
            <firmware_commit>b120233</firmware_commit>
        </cpn>
        <cpn id="31">
            <bus>1</bus>
            <board_name>servo_.MKS-HBL6625:010</board_name>
            <firmware_commit>b120233</firmware_commit>
        </cpn>
        <cpn id="32">
            <bus>1</bus>
            <board_name>servo_.MKS-HBL6625:006</board_name>
            <firmware_commit>b120233</firmware_commit>
        </cpn>
        <cpn id="33">
            <bus>1</bus>
            <board_name>servo_.MKS-HBL6625:011</board_name>
            <firmware_commit>b120233</firmware_commit>
        </cpn>
        <cpn id="34">
            <bus>1</bus>
            <board_name>gps_org.ardupilot.HolybroG4_GPS</board_name>
            <firmware_commit>6e519848</firmware_commit>
        </cpn>
        <cpn id="44">
            <bus>2</bus>
            <board_name>gps_org.ardupilot.HolybroG4_GPS</board_name>
            <firmware_commit>6e519848</firmware_commit>
        </cpn>
        <cpn id="35">
            <bus>1</bus>
            <board_name>CPN Fuse?</board_name>
            <firmware_commit>?</firmware_commit>
        </cpn>
        <cpn id="45">
            <bus>2</bus>
            <board_name>CPN Fuse?</board_name>
            <firmware_commit>?</firmware_commit>
        </cpn>
        <cpn id="36">
            <bus>1</bus>
            <board_name>arspd_org.ardupilot.f405_MatekGPS</board_name>
            <firmware_commit>158c7c49</firmware_commit>
        </cpn>
        <cpn id="46">
            <bus>2</bus>
            <board_name>arspd_org.ardupilot.f405_MatekGPS</board_name>
            <firmware_commit>158c7c49</firmware_commit>
        </cpn>
        <cpn id="38">
            <bus>1</bus>
            <board_name>batt_mon_io.p-systems.power-monitor</board_name>
            <firmware_commit>108b4bb7</firmware_commit>
        </cpn>
        <cpn id="39">
            <bus>1</bus>
            <board_name>fuel level?</board_name>
            <firmware_commit>108b4bb7</firmware_commit>
        </cpn>
        <cpn id="105">
            <bus>2</bus>
            <board_name>adsb_org.ardupilot.MatekL431-ADSB</board_name>
            <firmware_commit>9c6f307f</firmware_commit>
        </cpn>
        <cpn id="45">
            <bus>2</bus>
            <board_name>EIB??</board_name>
            <firmware_commit>?</firmware_commit>
        </cpn>
    </cpn_list>
    <esc_list>
    </esc_list>
    <airspeed_list>
    </airspeed_list>
    <battery_list>
    </battery_list>
    <barometer_list>
    </barometer_list>
    <lidar_list>
    </lidar_list>
    <lua_script_list>
        <lua_script>
            <source_path>libraries/AP_HAL_ChibiOS/hwdef/CarbonixCommon/scripts/cx_built_in_test.lua</source_path>
            <destination_path>cx_built_in_test.lua</destination_path>
        </lua_script>
        <lua_script>
            <source_path>libraries/AP_HAL_ChibiOS/hwdef/CarbonixCommon/scripts/cx_led_driver.lua</source_path>
            <destination_path>cx_led_driver.lua</destination_path>
        </lua_script>
        <lua_script>
            <source_path>libraries/AP_HAL_ChibiOS/hwdef/CarbonixCommon/scripts/cx_vtol_info.lua</source_path>
            <destination_path>cx_vtol_info.lua</destination_path>
        </lua_script>
        <lua_script>
            <source_path>libraries/AP_HAL_ChibiOS/hwdef/CarbonixCommon/scripts/modules/bit_esc.lua</source_path>
            <destination_path>modules/bit_esc.lua</destination_path>
        </lua_script>
        <lua_script>
            <source_path>libraries/AP_HAL_ChibiOS/hwdef/CarbonixCommon/scripts/modules/bit_gps.lua</source_path>
            <destination_path>modules/bit_gps.lua</destination_path>
        </lua_script>
        <lua_script>
            <source_path>libraries/AP_HAL_ChibiOS/hwdef/CarbonixCommon/scripts/modules/msg.lua</source_path>
            <destination_path>modules/msg.lua</destination_path>
        </lua_script>
        <lua_script>
            <source_path>libraries/AP_HAL_ChibiOS/hwdef/CarbonixCommon/scripts/aircraft/ottano.lua</source_path>
            <destination_path>modules/aircraft.lua</destination_path>
        </lua_script>
    </lua_script_list>
</data>

```

--baseline to start with will review/ test it before merge.

robertlong13 replied:
Ha, you left a review comment pending like an absolute noob

robertlong13 committed:
hwdef: add MatekL431-Carbonix-EIB

robertlong13 committed:
Tools: aircraft_config: delete board name prefixes

Unless we override the board name, the default appends org.ardupilot. to
it. We strip this out when looking for the firmware binary.

robertlong13 committed:
.github: build other non-CPN periph boards

robertlong13 committed:
CarbonixCommon: update script CRCs

robertlong13 commented:
[ParamsBeforeAfter.zip](https://github.com/user-attachments/files/20808660/ParamsBeforeAfter.zip)

I have reviewed the param changes for Volanti and older Ottano, and they are all good. Volanti got a couple non-functional-change option bits set and the script CRC updated (was out-of-date from a previous PR)

loki077 reviewed (approved):
LGTM.
Awesome work @robertlong13 !
